domain-http/interface/src/main/java/org/jboss/as/domain/http/server/StaticHeadersHandler.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2019 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jboss.as.domain.http.server;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.TreeMap;
+
+import io.undertow.server.HttpHandler;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.util.HttpString;
+
+/**
+ * An Undertow {@link HttpHandler} that can take a set of static headers to be applied to requests against a matched path.
+ *
+ * @author <a href="mailto:darran.lofthouse@jboss.com">Darran Lofthouse</a>
+ */
+class StaticHeadersHandler implements HttpHandler {
+
+    private final Map<String, List<HeaderConstant>> headersMap = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+
+    private final HttpHandler next;
+
+    StaticHeadersHandler(final HttpHandler next) {
+        this.next = next;
+    }
+
+    @Override
+    public void handleRequest(HttpServerExchange exchange) throws Exception {
+        String requestPath = exchange.getRelativePath();
+        for (Entry<String, List<HeaderConstant>> entry : headersMap.entrySet()) {
+            if (requestPath.startsWith(entry.getKey())) {
+                for (HeaderConstant constant : entry.getValue()) {
+                    constant.apply(exchange);
+                }
+            }
+        }
+
+        next.handleRequest(exchange);
+    }
+
+    void addHeader(final String path, final String header, final String value) {
+        final String actualPath = path.startsWith("/") ? path : "/" + path;
+        List<HeaderConstant> headers = headersMap.computeIfAbsent(actualPath, (s) -> new ArrayList<HeaderConstant>());
+        headers.add(new HeaderConstant(header, value));
+    }
+
+    private static class HeaderConstant {
+
+        private final HttpString headerName;
+        private final String value;
+
+        HeaderConstant(final String headerName, final String value) {
+            this.headerName = new HttpString(headerName);
+            this.value = value;
+        }
+
+        void apply(HttpServerExchange exchange) {
+            exchange.getResponseHeaders().add(headerName, value);
```

Comment:
@darranl According to the dev analysis, [Configured headers will not override any headers added by the corresponding endpoint](https://github.com/wildfly/wildfly-proposals/pull/263/files#diff-7135813aa89d8b31636212899b87ba26R63)

We should check here that the response headers does not already contain the header.

Reply:
I will be rewording that analysis section a little further after some debugging one of the WIPs today but in general that kind of check is not likely possible.

I have to set the headers as the request is on the way in as by the time it is on it's way out a handler could have already initiated sending the request to the client meaning I have no opportunity on the way out.

So for that sentence what I was really trying to say was we can set a header on the way in but as we don't know which headers an endpoint may also set we can not override that behaviour.